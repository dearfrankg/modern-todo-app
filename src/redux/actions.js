import * as types from 'redux/types';

export const addTodo = text => ({ type: types.ADD_TODO, text });
export const deleteTodo = id => ({ type: types.DELETE_TODO, id });
export const editTodo = (id, text) => ({ type: types.EDIT_TODO, id, text });
export const toggleTodo = todo => ({ type: types.TOGGLE_TODO, todo });
export const completeAll = () => ({ type: types.COMPLETE_ALL });
export const clearCompleted = () => ({ type: types.CLEAR_COMPLETED });
