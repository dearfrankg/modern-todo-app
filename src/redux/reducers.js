import * as types from 'redux/types';

export const initialState = {
  todos: [],
};

const todos = (state = initialState.todos, action) => {
  if (!action) return false;

  switch (action.type) {
    case types.ADD_TODO: {
      const maxId =
        state.length === 0
          ? 0
          : state.reduce((prev, current) => (prev.id > current.id ? prev : current)).id;

      return [
        {
          id: maxId + 1,
          text: action.text,
          completed: false,
        },
        ...state,
      ];
    }

    case types.DELETE_TODO: {
      return state.filter(todo => todo.id !== action.id);
    }

    case types.EDIT_TODO: {
      return state.map((todo) => {
        if (todo.id === action.id) return { ...todo, text: action.text };

        return todo;
      });
    }

    case types.TOGGLE_TODO: {
      return state.map((todo) => {
        if (todo.id === action.todo.id) return { ...todo, completed: !action.todo.completed };

        return todo;
      });
    }

    case types.COMPLETE_ALL: {
      const areAllMarked = state.every(todo => todo.completed);
      return state.map(todo => ({
        ...todo,
        completed: !areAllMarked,
      }));
    }

    case types.CLEAR_COMPLETED: {
      return state.filter(todo => todo.completed === false);
    }

    default:
      return state;
  }
};

export default todos;
