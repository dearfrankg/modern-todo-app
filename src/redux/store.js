import { createStore, combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

import rootReducer from 'redux/reducers';
import { loadState, saveState } from 'redux/localStorage';

const persistedState = loadState();

const store = createStore(
  combineReducers({
    todos: rootReducer,
    router: routerReducer,
  }),
  persistedState,
);

store.subscribe(() => {
  saveState({
    todos: store.getState().todos,
  });
});

export default store;
