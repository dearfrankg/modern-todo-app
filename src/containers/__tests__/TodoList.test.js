import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import { initialState } from 'redux/reducers';
import * as filters from 'redux/types';
import TodoListContainer, {
  TODO_FILTERS,
  showAll,
  showCompleted,
  showActive,
} from 'containers/TodoList';

const middlewares = [];
const mockStore = configureStore(middlewares);

const state = {
  router: {
    location: {
      hash: '#/',
    },
  },
  todos: initialState.todos,
};
const store = mockStore(state);

describe('TodoList Container:', () => {
  it('should render correctly with default props', () => {
    expect(
      shallow(
        <Provider store={store}>
          <TodoListContainer />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly with default filter', () => {
    expect(
      mount(
        <Provider store={store}>
          <TodoListContainer />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly with default filter but hash empty[2]', () => {
    const localState = {
      router: {
        location: {
          hash: '',
        },
      },
      todos: initialState.todos,
    };
    const localStore = mockStore(localState);

    expect(
      mount(
        <Provider store={localStore}>
          <TodoListContainer />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly with SHOW_ACTIVE filter', () => {
    const localState = {
      router: {
        location: {
          hash: '#/active',
        },
      },
      todos: initialState.todos,
    };
    const localStore = mockStore(localState);

    expect(
      mount(
        <Provider store={localStore}>
          <TodoListContainer />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly with SHOW_COMPLETED filter', () => {
    const localState = {
      router: {
        location: {
          hash: '#/completed',
        },
      },
      todos: initialState.todos,
    };
    const localStore = mockStore(localState);

    expect(
      mount(
        <Provider store={localStore}>
          <TodoListContainer />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  it('should render correctly when filter selected and some todos', () => {
    const localState = {
      router: {
        location: {
          hash: '#/completed',
        },
      },
      todos: [{ id: 1, text: 'text', completed: true }],
    };
    const localStore = mockStore(localState);

    expect(
      mount(
        <Provider store={localStore}>
          <TodoListContainer />
        </Provider>,
      ),
    ).toMatchSnapshot();
  });

  describe('TODO_FILTERS:', () => {
    it('should return correctly with SHOW_ALL', () => {
      expect(TODO_FILTERS[filters.SHOW_ALL]).toBe(showAll);
      expect(showAll()).toBe(true);
    });

    it('should return correctly with SHOW_COMPLETED', () => {
      expect(TODO_FILTERS[filters.SHOW_COMPLETED]).toBe(showCompleted);
      expect(showCompleted({ completed: true })).toBe(true);
    });

    it('should return correctly with SHOW_ACTIVE', () => {
      expect(TODO_FILTERS[filters.SHOW_ACTIVE]).toBe(showActive);
      expect(showActive({ completed: false })).toBe(true);
    });
  });
});
