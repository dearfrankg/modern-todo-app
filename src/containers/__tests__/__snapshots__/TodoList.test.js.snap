// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Container: should render correctly when filter selected and some todos 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(decoreWithStore)>
    <decoreWithStore
      dispatch={[Function]}
      router={
        Object {
          "location": Object {
            "hash": "#/completed",
          },
        }
      }
      todos={
        Array [
          Object {
            "completed": true,
            "id": 1,
            "text": "text",
          },
        ]
      }
    >
      <TodoListContainer
        dispatch={[Function]}
        router={
          Object {
            "location": Object {
              "hash": "#/completed",
            },
          }
        }
        todos={
          Array [
            Object {
              "completed": true,
              "id": 1,
              "text": "text",
            },
          ]
        }
      >
        <ul
          className="todo-list"
        >
          <Connect(TodoItemContainer)
            key="1"
            todo={
              Object {
                "completed": true,
                "id": 1,
                "text": "text",
              }
            }
          >
            <TodoItemContainer
              deleteTodo={[Function]}
              editTodo={[Function]}
              todo={
                Object {
                  "completed": true,
                  "id": 1,
                  "text": "text",
                }
              }
              toggleTodo={[Function]}
            >
              <TodoItem
                deleteTodo={[Function]}
                editText=""
                editing={false}
                handleChange={[Function]}
                handleEdit={[Function]}
                handleKeyDown={[Function]}
                handleSubmit={[Function]}
                inputRef={[Function]}
                key="1"
                todo={
                  Object {
                    "completed": true,
                    "id": 1,
                    "text": "text",
                  }
                }
                toggleTodo={[Function]}
              >
                <li
                  className="completed"
                  key="1"
                >
                  <div
                    className="view"
                  >
                    <input
                      checked={true}
                      className="toggle"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <label
                      onDoubleClick={[Function]}
                    >
                      text
                    </label>
                    <button
                      className="destroy"
                      onClick={[Function]}
                    />
                  </div>
                  <input
                    className="edit"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    value=""
                  />
                </li>
              </TodoItem>
            </TodoItemContainer>
          </Connect(TodoItemContainer)>
        </ul>
      </TodoListContainer>
    </decoreWithStore>
  </Connect(decoreWithStore)>
</Provider>
`;

exports[`TodoList Container: should render correctly with SHOW_ACTIVE filter 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(decoreWithStore)>
    <decoreWithStore
      dispatch={[Function]}
      router={
        Object {
          "location": Object {
            "hash": "#/active",
          },
        }
      }
      todos={Array []}
    >
      <TodoListContainer
        dispatch={[Function]}
        router={
          Object {
            "location": Object {
              "hash": "#/active",
            },
          }
        }
        todos={Array []}
      >
        <ul
          className="todo-list"
        />
      </TodoListContainer>
    </decoreWithStore>
  </Connect(decoreWithStore)>
</Provider>
`;

exports[`TodoList Container: should render correctly with SHOW_COMPLETED filter 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(decoreWithStore)>
    <decoreWithStore
      dispatch={[Function]}
      router={
        Object {
          "location": Object {
            "hash": "#/completed",
          },
        }
      }
      todos={Array []}
    >
      <TodoListContainer
        dispatch={[Function]}
        router={
          Object {
            "location": Object {
              "hash": "#/completed",
            },
          }
        }
        todos={Array []}
      >
        <ul
          className="todo-list"
        />
      </TodoListContainer>
    </decoreWithStore>
  </Connect(decoreWithStore)>
</Provider>
`;

exports[`TodoList Container: should render correctly with default filter 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(decoreWithStore)>
    <decoreWithStore
      dispatch={[Function]}
      router={
        Object {
          "location": Object {
            "hash": "#/",
          },
        }
      }
      todos={Array []}
    >
      <TodoListContainer
        dispatch={[Function]}
        router={
          Object {
            "location": Object {
              "hash": "#/",
            },
          }
        }
        todos={Array []}
      >
        <ul
          className="todo-list"
        />
      </TodoListContainer>
    </decoreWithStore>
  </Connect(decoreWithStore)>
</Provider>
`;

exports[`TodoList Container: should render correctly with default filter but hash empty[2] 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(decoreWithStore)>
    <decoreWithStore
      dispatch={[Function]}
      router={
        Object {
          "location": Object {
            "hash": "",
          },
        }
      }
      todos={Array []}
    >
      <TodoListContainer
        dispatch={[Function]}
        router={
          Object {
            "location": Object {
              "hash": "",
            },
          }
        }
        todos={Array []}
      >
        <ul
          className="todo-list"
        />
      </TodoListContainer>
    </decoreWithStore>
  </Connect(decoreWithStore)>
</Provider>
`;

exports[`TodoList Container: should render correctly with default props 1`] = `<Connect(decoreWithStore) />`;
